### PR TITLE
update-plugin-list: use json, use http caching, use gh action caches

### PR DESCRIPTION
**Pull request by RonnyPfannschmidt**
this change
* switches from html index usage to json index usage
* adds using requests-cache as http cache to trim down runtime
* adds gh action caches for pip and the new http cache

### CODE REVIEW

File: .github/workflows/update-plugin-list.yml [hunk 1]
```diff
@@ -27,12 +27,20 @@ jobs:
...
         with:
-          python-version: 3.8
+          python-version: "3.8"
+          cache: pip
+      - name: requests-cache
+        uses: actions/cache@v3
+        with:
+          path: ~/.cache/pytest-plugin-list/
+          key: plugins-http-cache-${{ github.run_id }} # Can use time based key as well
+          restore-keys: plugins-http-cache-
```
Comment (bluetech) [inline .github/workflows/update-plugin-list.yml:37]: What's the reason to add `github.run_id` to the cache key instead of using a constant key?

Comment (RonnyPfannschmidt) [reply .github/workflows/update-plugin-list.yml:37]: following https://github.com/actions/cache/blob/main/tips-and-workarounds.md#update-a-cache
a cachi is immutable, so the new version needs a new key

**Review by bluetech — APPROVED** [review]

**Comment by nicoddemus** [comment]
This is interesting, but I wonder if we need to speed up this workflow at all? This runs only once a week, and [looking at the list of runs](https://github.com/pytest-dev/pytest/actions/workflows/update-plugin-list.yml), each run takes only around 4m or less.

**Comment by RonnyPfannschmidt** [comment]
When I hacked on it to use json, the feedback time was a pain, adding the cache made it a breeze to run

**Comment by nicoddemus** [comment]
Got it, thanks!

**Review by nicoddemus — APPROVED** [review]

File: doc/en/reference/plugin_list.rst [hunk 2]
```diff
@@ -227,7 +229,7 @@ This list contains 1243 plugins.
...
    :pypi:`pytest-cpp`                               Use pytest's runner to discover and execute C++ tests                                                                                                                                                     Jan 30, 2023    5 - Production/Stable  pytest (>=7.0)
-   :pypi:`pytest-cppython`                          A pytest plugin that imports CPPython testing types                                                                                                                                                       Apr 20, 2023    N/A                    N/A
+   :pypi:`pytest-cppython`                          A pytest plugin that imports CPPython testing types                                                                                                                                                       Mar 20, 2023    N/A                    N/A
```
Comment (RonnyPfannschmidt) [inline doc/en/reference/plugin_list.rst:232]: there appears to be a problem with the returned data as the dates change

**Comment by RonnyPfannschmidt** [comment]
fixing the false cache-hit issue wrt last-serial made it more complicated, lets reeval

**Review by bluetech — APPROVED** [review]
> fixing the false cache-hit issue wrt last-serial made it more complicated, lets reeval

Didn't realize requests-cache ignores the cache-control header by default. The fix LGTM.

**Comment by RonnyPfannschmidt** [comment]
@bluetech it's not that it ignores it, it's just that the default cache controls lead to stale data, one needs to compare extra data to force a refresh

None of the http standard caching tools enable the required data pass over, so manual intervention is needed

File: .github/workflows/update-plugin-list.yml [hunk 3]
```diff
@@ -27,12 +27,20 @@ jobs:
       - name: Setup Python
         uses: actions/setup-python@v4
         with:
-          python-version: 3.8
+          python-version: "3.8"
+          cache: pip
+      - name: requests-cache
```
Comment (nicoddemus) [inline .github/workflows/update-plugin-list.yml:32]: For consistent with the other step names.

```suggestion
      - name: Requests Cache
```

File: .github/workflows/update-plugin-list.yml [hunk 4]
```diff
@@ -27,12 +27,20 @@ jobs:
...
+        uses: actions/cache@v3
+        with:
+          path: ~/.cache/pytest-plugin-list/
+          key: plugins-http-cache-${{ github.run_id }} # Can use time based key as well
```
Comment (nicoddemus) [inline .github/workflows/update-plugin-list.yml:36]: Did you mean "could" here?

```suggestion
          key: plugins-http-cache-${{ github.run_id }} # Could use time based key as well.
```

**Review by nicoddemus — APPROVED** [review]

**Pull request merged**